### PR TITLE
Localize default messages

### LINKED ISSUE #1195: trouble with native compatibility with ajv-i18n
### Prerequisites

- [v] I have read the [documentation](https://react-jsonschema-form.readthedocs.io/);
- [v] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [v] I'm providing a playground **codesandbox** (https://codesandbox.io/embed/vm3lz56j53?fontsize=14) demonstrating the issue.

### Description

Current realization of errors object incomes in **transformErrors** (https://react-jsonschema-form.readthedocs.io/en/latest/validation/#custom-error-messages) is incompatible with **ajv-i18n**  (https://github.com/epoberezkin/ajv-i18n)  

Since the AJV is used internally, it would be nice to ensure compatibility with AJV native i18n gear, especially considering that such compatibility is very simple to provide by expanding the "erorrs" with one property "keyword", the value for which already exists in members of incoming errors array members as "name" property.

See my playground sample (https://codesandbox.io/embed/vm3lz56j53?fontsize=14), where basic erros array normalizing by map for full compilibility with ajv-i18n localization function.

### Steps to Reproduce

1. Install and import localization (as sample - directly ru) function for ajv-i18n npm package
2. declare transformErrors fucntion where try to call localization function on basic (non modified) errors incoming argument
`
function transformErrors(errors) {
	localize_ru(errors);
	return i18n_normalized;
}
`
3. You will take error about missing required 'keyword' property in ajv-i18n
4. Then modify your transformErrors, where you previously normilize errors object with basic map
`
function transformErrors(errors) {
	const i18n_normalized = errors.map(e => ({ ...e, keyword: e.name }));
	localize_ru(i18n_normalized);
	return i18n_normalized;
}
`
5. errors messages are succsefully localizated with ajv-i18n

#### Expected behavior

The erros array members (objects) will have 'keyword' property (duplicated 'name' property value)

PS: May be 'name' property will be renamed to 'keyword' with concurrent refactoring thats usages in other project sources

#### Actual behavior

The erros array members (objects) not have 'keyword' property, and ajv-i18n localization gear crashes called with the current errors array as argumnet

### Version

current major - 1.2.1

### PR DESCRIPTION
**Pull request by miguelcast**
### Localize default messages

Add localize prop to Form component for translate messages to different languages. Fixes #1195

### Checklist

* [x] **I'm updating documentation**
  - [ ] I've checked the rendering of the Markdown text I've added
  - [x] If I'm adding a new section, I've updated the Table of Content
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [x] I've updated docs if needed
  - [x] I've run `npm run cs-format` on my branch to conform my code to [prettier](https://github.com/prettier/prettier) coding style
* [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by miguelcast** [comment]
Why can the "run cs-check" action fail?

> Failed at the react-jsonschema-form@1.1.0 cs-check script 'prettier -l $npm_package_prettierOptions '{playground,src,test}/**/*.js''.

File: README.md [hunk 1]
```diff
@@ -1471,6 +1472,21 @@ One consequence of this is that if you have an empty string in your `enum` array
 
 If you want to have the field set to a default value when empty you can provide a `ui:emptyValue` field in the `uiSchema` object.
 
+## Localize default messages
+
+You can change localize to default messages errors.
```
Comment (edi9999) [inline README.md]: `You can change localize to default messages errors.`

=> 

`You can change the localization of the error messages.`

File: README.md [hunk 2]
```diff
@@ -1471,6 +1472,21 @@ One consequence of this is that if you have an empty string in your `enum` array
...
+ 
+Here are some examples from the [playground](http://mozilla-services.github.io/react-jsonschema-form/).
+ 
+Here list of languages support [ajv-i18n](https://github.com/epoberezkin/ajv-i18n).
```
Comment (edi9999) [inline README.md]: Here is the list of supported languages [ajv-i18n](https://github.com/epoberezkin/ajv-i18n).

File: README.md [hunk 3]
```diff
@@ -1471,6 +1472,21 @@ One consequence of this is that if you have an empty string in your `enum` array
...
+
+```jsx
+render((
+  <Form schema={schema}
+        localize="es" />
```
Comment (edi9999) [inline README.md]: localization="es"

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-jsonschema-form",
-  "version": "1.0.2",
+  "version": "1.1.2",
```
Comment (edi9999) [inline package.json]: please don't change version here.

**Comment by edi9999** [comment]
Can you add a test showing usage of this new feature ?

I would also like to think if we could add support for localization without adding a dependency to this repo (since it increases the bundle size for each user, despite not using every feature)

File: src/validate.js [hunk 5]
```diff
@@ -152,10 +153,12 @@ export default function validateFormData(
   formData,
   schema,
   customValidate,
-  transformErrors
+  transformErrors,
+  local
 ) {
   try {
     ajv.validate(schema, formData);
+    localize[local] && localize[local](ajv.errors);
```
Comment (pieplu) [inline src/validate.js]: All avji18n will be imported. Why, insted of take a "string" in props, taking a function, that can perform the localization ?
The user (dev?) will can import only needed local function, and pass it. 

```suggestion
    localize && localize(ajv.errors);
```

Comment (kopax) [reply src/validate.js]: I agree

File: test/validate_test.js [hunk 6]
```diff
@@ -294,21 +294,23 @@ describe("Validation", () => {
...
+          expect(comp.state.errors[0].message).eql(
+            "should have required property foo"
+          );
         });
 
         it("should render errors", () => {
           expect(node.querySelectorAll(".errors li")).to.have.length.of(1);
           expect(node.querySelector(".errors li").textContent).eql(
-            ".foo is a required property"
+            ".foo should have required property foo"
```
[the remark this answers is not in the document]
Comment (miguelcast) [reply test/validate_test.js]: I will remove this dependence

File: src/components/Form.js [hunk 7]
```diff
@@ -88,14 +88,15 @@ export default class Form extends Component {
...
+    const { validate, transformErrors, localization } = this.props;
     const { definitions } = this.getRegistry();
     const resolvedSchema = retrieveSchema(schema, definitions, formData);
     return validateFormData(
       formData,
       resolvedSchema,
       validate,
-      transformErrors
+      transformErrors,
+      localization
```
Comment (epicfaace) [inline src/components/Form.js]: Why are you returning `localization` from `validateFormData` (i.e. what is the use case for doing so)?

File: src/components/Form.js [hunk 8]
```diff
@@ -294,5 +295,6 @@ if (process.env.NODE_ENV !== "production") {
     transformErrors: PropTypes.func,
     safeRenderCompletion: PropTypes.bool,
     formContext: PropTypes.object,
+    localization: PropTypes.func,
```
Comment (epicfaace) [inline src/components/Form.js]: Why did you rename it from `localize` to `localization`? Just want to hear your thinking on this.

File: src/validate.js [hunk 9]
```diff
@@ -127,7 +127,6 @@ function transformAjvErrors(errors = []) {
   if (errors === null) {
     return [];
   }
```
Comment (epicfaace) [inline src/validate.js]:
```suggestion
   }
  
```

File: playground/samples/localization.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+const messages = {
```
Comment (epicfaace) [inline playground/samples/localization.js]: This is fine, but you might as well just import a localize function from `ajv-i18n` rather than writing your own `convertMessagesText` function for this playground example -- I'm guessing that's simpler? I'd assume that most of the use cases from localization would use an existing function corresponding to a language from `ajv-i18n`.

**Comment by epicfaace** [comment]
@miguelcast do you have time to respond to my comments and make the requested changes on this?

**Comment by miguelcast** [comment]
> @miguelcast do you have time to respond to my comments and make the requested changes on this?

Excuse me for the delay, I thinking finish this feature these days.

**Comment by epicfaace** [comment]
@miguelcast just bumping this!

**Comment by miguelcast** [comment]
> @miguelcast just bumping this!

Hi, There are 2 tests with error but I don't know how to solve it and neither who inserted these?

**Comment by Maxhy** [comment]
@miguelcast any chance to finish this or forward this stuff to someone else? A lot of localized features depends on it and as there is already a pending pull request, we are all waiting for it haha. Thanks!

**Comment by epicfaace** [comment]
Anyone is welcome to pick this up if they would like this feature to be added.

**Comment by epicfaace** [comment]
Closing this for now due to inactivity, though this PR could be a good starting point for a later implementation of localization.

**Comment by IgorLesnevskiy** [comment]
@epicfaace what's left to finish here? Guess I could do it.

**Comment by miguelcast** [comment]
> @epicfaace what's left to finish here? Guess I could do it.

The issue is long overdue, you have to update the PR and fix the unit tests to be able to finish it.
